crates/red_knot_python_semantic/src/types/context.rs

```diff
@@ -0,0 +1,131 @@
+use std::fmt;
+
+use drop_bomb::DropBomb;
+use ruff_db::{
+    diagnostic::{DiagnosticId, Severity},
+    files::File,
+};
+use ruff_python_ast::AnyNodeRef;
+use ruff_text_size::Ranged;
+
+use crate::{
+    lint::{LintId, LintMetadata},
+    Db,
+};
+
+use super::{TypeCheckDiagnostic, TypeCheckDiagnostics};
+
+/// Context for inferring the types of a single file.
+///
+/// One context exists for at least for every inferred region but it's
+/// possible that inferring a sub-region, like an unpack assignment, creates
+/// a sub-context.
+///
+/// Tracks the reported diagnostics of the inferred region.
+///
+/// ## Consuming
+/// It's important that the context is explicitly consumed before dropping by calling
+/// [`InferContext::finish`] and the returned diagnostics must be stored
+/// on the current [`TypeInference`](super::infer::TypeInference) result.
+pub(crate) struct InferContext<'db> {
+    db: &'db dyn Db,
+    file: File,
+    diagnostics: std::cell::RefCell<TypeCheckDiagnostics>,
+    defused: DropBomb,
```

Comment:
minor nit: the name `defused` confused me at first, because initially this is not defused! It's only defused after `finish()`. Would have been clearer to me if it were named `bomb` or `drop_bomb`